### PR TITLE
better initialization

### PR DESCRIPTION
**Pull request by StrongMonkey**

### CODE REVIEW

**Comment by StrongMonkey** [comment]
https://github.com/rancher/rancher/issues/8639

File: events/handler.go [hunk 1]
```diff
@@ -54,6 +54,17 @@ func (h *schedulingHandler) Prioritize(event *revents.Event, client *client.Ranc
...
+		err = h.scheduler.UpdateWithMetadata(false)
+		if err != nil {
+			if i == 4 {
+				panic(fmt.Sprintf("Failed at metadata initialization. 5 consecutive errors attempting to reach metadata. Panicing. Error: %v", err))
+			}
+			continue
+		}
+		break
+	}
+
```
Comment (StrongMonkey) [inline events/handler.go:67]: we always call `UpdateWithMetadata` here before we dive into the real priortize action. The error is always going to be a metadata error.

File: main.go [hunk 2]
```diff
@@ -93,6 +94,14 @@ func run(c *cli.Context) error {
...
 	}(exit)
 
+	go func() {
+		for {
+			t.Sleep(t.Minute * 3)
+			logrus.Info("Locking scheduler now. Sync scheduler information with rancher metadata")
+			scheduler.UpdateWithMetadata(true)
+		}
+	}()
+
```
Comment (StrongMonkey) [inline main.go:104]: Sync with metadata every 3 minutes to make sure everything is updated.

File: scheduler/scheduler.go [hunk 3]
```diff
@@ -172,6 +190,188 @@ func (s *Scheduler) CompareHostLabels(hosts []metadata.Host) bool {
...
+		}
+
+		usedRes[instancePool]++
+		usedRes[memoryPool] += c.MemoryReservation
+		usedRes[cpuPool] += c.MilliCPUReservation
+	}
+
+	return resourcesByHost, nil
+}
+
```
Comment (StrongMonkey) [inline scheduler/scheduler.go]: duplicate code. May need to refactor later on.

**Comment by StrongMonkey** [comment]
@cjellick Can you plz take a look at this pr to see if this meets your initial design?

**Comment by cjellick** [comment]
We can't merge this until we know a couple of things:
1. How it performs with many hosts and containers (1000 containers spread across 10 hosts would be a good start). Our metadata sync code is deifnitely inefficient and makes a lot of calls. So, I want to know generally how long that takes.
2. I need to know if this has the potential to make things worse in some circumstances. Here is the the scenario I'm concerned about: immediately after scheduling a bunch of containers, the sync with metadata kicks in and wipes out the containers that have been scheduled but are not yet in MD.
  - probably need @ibuildthecloud's input on a way around this. Maybe we only do the MD resync if there is been a quiet period of 15 (or N) seconds where no scheduling events have come in.

**Comment by StrongMonkey** [comment]
@cjellick It takes 2.4739 seconds for syncing the metadata(1000 containers with 10 hosts)
```
5/24/2017 11:24:01 AMtime="2017-05-24T18:24:01Z" level=info msg="Locking scheduler now. Sync scheduler information with rancher metadata"
5/24/2017 11:24:01 AMtime="2017-05-24T18:24:01Z" level=info msg="Updating resource pool [storageSize] with total 39916912 and used 0 for host  071d10b6-42b6-4a44-ad27-b9af15313660"
5/24/2017 11:24:01 AMtime="2017-05-24T18:24:01Z" level=info msg="Updating resource pool [instanceReservation] with total 1000000 and used 108 for host  071d10b6-42b6-4a44-ad27-b9af15313660"
5/24/2017 11:24:01 AMtime="2017-05-24T18:24:01Z" level=info msg="Updating resource pool [cpuReservation] with total 1000 and used 0 for host  071d10b6-42b6-4a44-ad27-b9af15313660"
5/24/2017 11:24:01 AMtime="2017-05-24T18:24:01Z" level=info msg="Updating resource pool [memoryReservation] with total 1777336320 and used 0 for host  071d10b6-42b6-4a44-ad27-b9af15313660"
5/24/2017 11:24:01 AMtime="2017-05-24T18:24:01Z" level=info msg="Updating resource pool [portReservation], ip set [0.0.0.0], ports map tcp map[0.0.0.0:map[]], ports map udp map[0.0.0.0:map[500:df75fbff-00f7-4f07-a9b1-b8219ddf1898 4500:df75fbff-00f7-4f07-a9b1-b8219ddf1898]] for host 071d10b6-42b6-4a44-ad27-b9af15313660"
5/24/2017 11:24:01 AMtime="2017-05-24T18:24:01Z" level=info msg="Updating resource pool [storageSize] with total 39916908 and used 0 for host  20bd80bb-ff8c-4de6-8f1f-4e8ff981f48b"
5/24/2017 11:24:01 AMtime="2017-05-24T18:24:01Z" level=info msg="Updating resource pool [instanceReservation] with total 1000000 and used 108 for host  20bd80bb-ff8c-4de6-8f1f-4e8ff981f48b"
5/24/2017 11:24:01 AMtime="2017-05-24T18:24:01Z" level=info msg="Updating resource pool [cpuReservation] with total 1000 and used 0 for host  20bd80bb-ff8c-4de6-8f1f-4e8ff981f48b"
5/24/2017 11:24:01 AMtime="2017-05-24T18:24:01Z" level=info msg="Updating resource pool [memoryReservation] with total 1777336320 and used 0 for host  20bd80bb-ff8c-4de6-8f1f-4e8ff981f48b"
5/24/2017 11:24:01 AMtime="2017-05-24T18:24:01Z" level=info msg="Updating resource pool [portReservation], ip set [0.0.0.0], ports map tcp map[0.0.0.0:map[]], ports map udp map[0.0.0.0:map[500:182fc10d-a23a-42d1-a9b9-215752ea9ae9 4500:182fc10d-a23a-42d1-a9b9-215752ea9ae9]] for host 20bd80bb-ff8c-4de6-8f1f-4e8ff981f48b"
5/24/2017 11:24:01 AMtime="2017-05-24T18:24:01Z" level=info msg="Updating resource pool [instanceReservation] with total 1000000 and used 108 for host  07a30140-4157-4103-a382-7e28a2391d8a"
5/24/2017 11:24:01 AMtime="2017-05-24T18:24:01Z" level=info msg="Updating resource pool [cpuReservation] with total 1000 and used 0 for host  07a30140-4157-4103-a382-7e28a2391d8a"
5/24/2017 11:24:01 AMtime="2017-05-24T18:24:01Z" level=info msg="Updating resource pool [memoryReservation] with total 1777336320 and used 0 for host  07a30140-4157-4103-a382-7e28a2391d8a"
5/24/2017 11:24:01 AMtime="2017-05-24T18:24:01Z" level=info msg="Updating resource pool [storageSize] with total 39916912 and used 0 for host  07a30140-4157-4103-a382-7e28a2391d8a"
5/24/2017 11:24:01 AMtime="2017-05-24T18:24:01Z" level=info msg="Updating resource pool [portReservation], ip set [0.0.0.0], ports map tcp map[0.0.0.0:map[]], ports map udp map[0.0.0.0:map[500:fdc88e60-372f-4de0-ba39-bb0acd4f1e6d 4500:fdc88e60-372f-4de0-ba39-bb0acd4f1e6d]] for host 07a30140-4157-4103-a382-7e28a2391d8a"
5/24/2017 11:24:01 AMtime="2017-05-24T18:24:01Z" level=info msg="Updating resource pool [instanceReservation] with total 1000000 and used 108 for host  dc1a0e3f-1f11-4cea-8d76-72729cf40eb8"
5/24/2017 11:24:01 AMtime="2017-05-24T18:24:01Z" level=info msg="Updating resource pool [cpuReservation] with total 1000 and used 0 for host  dc1a0e3f-1f11-4cea-8d76-72729cf40eb8"
5/24/2017 11:24:01 AMtime="2017-05-24T18:24:01Z" level=info msg="Updating resource pool [memoryReservation] with total 1777336320 and used 0 for host  dc1a0e3f-1f11-4cea-8d76-72729cf40eb8"
5/24/2017 11:24:01 AMtime="2017-05-24T18:24:01Z" level=info msg="Updating resource pool [storageSize] with total 39916908 and used 0 for host  dc1a0e3f-1f11-4cea-8d76-72729cf40eb8"
5/24/2017 11:24:02 AMtime="2017-05-24T18:24:02Z" level=info msg="Updating resource pool [portReservation], ip set [0.0.0.0], ports map tcp map[0.0.0.0:map[]], ports map udp map[0.0.0.0:map[500:dbb1ed4d-eeac-4d82-9800-8e4570310103 4500:dbb1ed4d-eeac-4d82-9800-8e4570310103]] for host dc1a0e3f-1f11-4cea-8d76-72729cf40eb8"
5/24/2017 11:24:02 AMtime="2017-05-24T18:24:02Z" level=info msg="Updating resource pool [instanceReservation] with total 1000000 and used 108 for host  aa5dd320-9ee8-4d7a-a89f-faeb422a7434"
5/24/2017 11:24:02 AMtime="2017-05-24T18:24:02Z" level=info msg="Updating resource pool [cpuReservation] with total 1000 and used 0 for host  aa5dd320-9ee8-4d7a-a89f-faeb422a7434"
5/24/2017 11:24:02 AMtime="2017-05-24T18:24:02Z" level=info msg="Updating resource pool [memoryReservation] with total 1777336320 and used 0 for host  aa5dd320-9ee8-4d7a-a89f-faeb422a7434"
5/24/2017 11:24:02 AMtime="2017-05-24T18:24:02Z" level=info msg="Updating resource pool [storageSize] with total 39916908 and used 0 for host  aa5dd320-9ee8-4d7a-a89f-faeb422a7434"
5/24/2017 11:24:02 AMtime="2017-05-24T18:24:02Z" level=info msg="Updating resource pool [portReservation], ip set [0.0.0.0], ports map tcp map[0.0.0.0:map[]], ports map udp map[0.0.0.0:map[500:c86724fc-8f37-4ee3-b63e-23c9f730ffc7 4500:c86724fc-8f37-4ee3-b63e-23c9f730ffc7]] for host aa5dd320-9ee8-4d7a-a89f-faeb422a7434"
5/24/2017 11:24:02 AMtime="2017-05-24T18:24:02Z" level=info msg="Updating resource pool [instanceReservation] with total 1000000 and used 108 for host  fdd82014-f569-4778-9503-55a2bff897ac"
5/24/2017 11:24:02 AMtime="2017-05-24T18:24:02Z" level=info msg="Updating resource pool [cpuReservation] with total 1000 and used 0 for host  fdd82014-f569-4778-9503-55a2bff897ac"
5/24/2017 11:24:02 AMtime="2017-05-24T18:24:02Z" level=info msg="Updating resource pool [memoryReservation] with total 1777336320 and used 0 for host  fdd82014-f569-4778-9503-55a2bff897ac"
5/24/2017 11:24:02 AMtime="2017-05-24T18:24:02Z" level=info msg="Updating resource pool [storageSize] with total 39916904 and used 0 for host  fdd82014-f569-4778-9503-55a2bff897ac"
5/24/2017 11:24:02 AMtime="2017-05-24T18:24:02Z" level=info msg="Updating resource pool [portReservation], ip set [0.0.0.0], ports map tcp map[0.0.0.0:map[]], ports map udp map[0.0.0.0:map[4500:d143f00c-be59-40e5-98de-c31f47ef4f87 500:d143f00c-be59-40e5-98de-c31f47ef4f87]] for host fdd82014-f569-4778-9503-55a2bff897ac"
5/24/2017 11:24:02 AMtime="2017-05-24T18:24:02Z" level=info msg="Updating resource pool [memoryReservation] with total 1777336320 and used 0 for host  2ed69790-57e7-482d-9be6-2476ec66dc80"
5/24/2017 11:24:02 AMtime="2017-05-24T18:24:02Z" level=info msg="Updating resource pool [storageSize] with total 39916912 and used 0 for host  2ed69790-57e7-482d-9be6-2476ec66dc80"
5/24/2017 11:24:02 AMtime="2017-05-24T18:24:02Z" level=info msg="Updating resource pool [instanceReservation] with total 1000000 and used 109 for host  2ed69790-57e7-482d-9be6-2476ec66dc80"
5/24/2017 11:24:02 AMtime="2017-05-24T18:24:02Z" level=info msg="Updating resource pool [cpuReservation] with total 1000 and used 0 for host  2ed69790-57e7-482d-9be6-2476ec66dc80"
5/24/2017 11:24:02 AMtime="2017-05-24T18:24:02Z" level=info msg="Updating resource pool [portReservation], ip set [0.0.0.0], ports map tcp map[0.0.0.0:map[]], ports map udp map[0.0.0.0:map[500:6643a2d5-e8e8-4d55-a140-5cbc7060e0f5 4500:6643a2d5-e8e8-4d55-a140-5cbc7060e0f5]] for host 2ed69790-57e7-482d-9be6-2476ec66dc80"
5/24/2017 11:24:02 AMtime="2017-05-24T18:24:02Z" level=info msg="Updating resource pool [instanceReservation] with total 1000000 and used 108 for host  84179394-3f63-4f84-aa5e-6b019746f6a4"
5/24/2017 11:24:02 AMtime="2017-05-24T18:24:02Z" level=info msg="Updating resource pool [cpuReservation] with total 1000 and used 0 for host  84179394-3f63-4f84-aa5e-6b019746f6a4"
5/24/2017 11:24:02 AMtime="2017-05-24T18:24:02Z" level=info msg="Updating resource pool [memoryReservation] with total 1777336320 and used 0 for host  84179394-3f63-4f84-aa5e-6b019746f6a4"
5/24/2017 11:24:02 AMtime="2017-05-24T18:24:02Z" level=info msg="Updating resource pool [storageSize] with total 39916904 and used 0 for host  84179394-3f63-4f84-aa5e-6b019746f6a4"
5/24/2017 11:24:03 AMtime="2017-05-24T18:24:03Z" level=info msg="Updating resource pool [portReservation], ip set [0.0.0.0], ports map tcp map[0.0.0.0:map[]], ports map udp map[0.0.0.0:map[500:46328163-9677-4110-8810-c1df1cc8bb01 4500:46328163-9677-4110-8810-c1df1cc8bb01]] for host 84179394-3f63-4f84-aa5e-6b019746f6a4"
5/24/2017 11:24:03 AMtime="2017-05-24T18:24:03Z" level=info msg="Updating resource pool [memoryReservation] with total 1777336320 and used 0 for host  744178fe-c741-4e7e-ad8e-4ff02aa16b48"
5/24/2017 11:24:03 AMtime="2017-05-24T18:24:03Z" level=info msg="Updating resource pool [storageSize] with total 39916908 and used 0 for host  744178fe-c741-4e7e-ad8e-4ff02aa16b48"
5/24/2017 11:24:03 AMtime="2017-05-24T18:24:03Z" level=info msg="Updating resource pool [instanceReservation] with total 1000000 and used 108 for host  744178fe-c741-4e7e-ad8e-4ff02aa16b48"
5/24/2017 11:24:03 AMtime="2017-05-24T18:24:03Z" level=info msg="Updating resource pool [cpuReservation] with total 1000 and used 0 for host  744178fe-c741-4e7e-ad8e-4ff02aa16b48"
5/24/2017 11:24:03 AMtime="2017-05-24T18:24:03Z" level=info msg="Updating resource pool [portReservation], ip set [0.0.0.0], ports map tcp map[0.0.0.0:map[]], ports map udp map[0.0.0.0:map[500:319620f4-7772-46ad-abbe-5a76a4786f96 4500:319620f4-7772-46ad-abbe-5a76a4786f96]] for host 744178fe-c741-4e7e-ad8e-4ff02aa16b48"
5/24/2017 11:24:03 AMtime="2017-05-24T18:24:03Z" level=info msg="Updating resource pool [storageSize] with total 39916908 and used 0 for host  fbc2d95a-775f-4297-8105-3e101c45b362"
5/24/2017 11:24:03 AMtime="2017-05-24T18:24:03Z" level=info msg="Updating resource pool [instanceReservation] with total 1000000 and used 108 for host  fbc2d95a-775f-4297-8105-3e101c45b362"
5/24/2017 11:24:03 AMtime="2017-05-24T18:24:03Z" level=info msg="Updating resource pool [cpuReservation] with total 1000 and used 0 for host  fbc2d95a-775f-4297-8105-3e101c45b362"
5/24/2017 11:24:03 AMtime="2017-05-24T18:24:03Z" level=info msg="Updating resource pool [memoryReservation] with total 1777336320 and used 0 for host  fbc2d95a-775f-4297-8105-3e101c45b362"
5/24/2017 11:24:03 AMtime="2017-05-24T18:24:03Z" level=info msg="Updating resource pool [portReservation], ip set [0.0.0.0], ports map tcp map[0.0.0.0:map[]], ports map udp map[0.0.0.0:map[500:24811349-0d6e-4492-b27a-e6cb7c24b2df 4500:24811349-0d6e-4492-b27a-e6cb7c24b2df]] for host fbc2d95a-775f-4297-8105-3e101c45b362"
5/24/2017 11:24:03 AMtime="2017-05-24T18:24:03Z" level=info msg=2.473915642
```
screen shot 2017-05-24 at 11 25 41 am

**Comment by StrongMonkey** [comment]
the time it takes really varies. But in average it takes 2s to sync.

**Comment by cjellick** [comment]
What does really vaires mean? whats the range? Did your test include ports because ports are pretty significant to the scheduler MD logic

File: events/handler.go [hunk 4]
```diff
@@ -54,6 +54,17 @@ func (h *schedulingHandler) Prioritize(event *revents.Event, client *client.Ranc
 		return errors.Wrapf(err, "Error decoding prioritize event %v.", event)
 	}
 
+	for i := 0; i < 5; i++ {
+		err = h.scheduler.UpdateWithMetadata(false)
+		if err != nil {
+			if i == 4 {
+				panic(fmt.Sprintf("Failed at metadata initialization. 5 consecutive errors attempting to reach metadata. Panicing. Error: %v", err))
```
Comment (cjellick) [inline events/handler.go:61]: Do you really need to panic? Can you buble up an error?

Comment (StrongMonkey) [reply events/handler.go:61]: https://github.com/rancher/scheduler/blob/master/resourcewatchers/metadata.go#L175 I followed the same pattern in here.

**Comment by StrongMonkey** [comment]
I didn't test with ports. But i think that will be the same result. I suspect most of the time spent on metadata.getHost() and metadata.getContainer(), regardless of it has ports or not. After data has been pulled from metadata, what's left are just in-memory operations which should be fast though. There are some possible optimizations to reduce the call for getContainers(). Range is between 1.5s to 2.5s.

**Comment by StrongMonkey** [comment]
btw, our scheduler did a perfect job balancing these 1000 containers. Each host has exactly 108 containers.(100 + 8 infra containers) 😄

**Comment by cjellick** [comment]
@StrongMonkey The port method does another call to client.GetContainers, but I guess it does it regardless of whether or not ports are used, so you're right in that it doesnt make a difference

**Pull request closed without merging**